### ISSUE TITLE
Weighted custom loss for pixelwise classification

### ISSUE THREAD

**Issue by benjamin-robbins** [keras-team/keras#8308]
https://github.com/fchollet/keras/issues/6261

I have tried the approach suggested in 6261 [](https://github.com/fchollet/keras/issues/6261) and am still coming up with poor results. I have an unbalanced data set of active/nonactive pixels that I need to be able to weight according to class. I am not able to oversample, generate data, etc that one would normally be able to do given unbalanced data.

The final layers of my U-net is:

conv10 = Reshape((self.num_class, self.img_rows * self.img_cols))(conv10)
conv10 = Permute((2, 1))(conv10)
softmax = Activation('softmax')(conv10)

This gives me a [rows*cols, 2] matrix where the probability for each pixel is each row - where the first column in the probability of that pixel being 0 and the second column is the probability of the pixel being one.

I one-hot my labels using keras's to_categorical function so that my label is also in the form of [row*col, 2]

I then pass weights such as [1,8] to 

`def weighted_pixelwise_crossentropy(class_weights):
    
    def loss(y_true, y_pred):
        epsilon = _to_tensor(_EPSILON, y_pred.dtype.base_dtype)
        y_pred = tf.clip_by_value(y_pred, epsilon, 1. - epsilon)
        return - tf.reduce_sum(tf.multiply(y_true * tf.log(y_pred), class_weights))

    return loss`

However, I get back results whereby all predictions are .5 for both classes no matter what values for the weights I put in.

Any pointers would be GREATLY appreciated. I have spent quite a bit of time on this with no luck.

Thanks!

**Comment by rex-yue-wu** [comment]
update your keras, the recent versions support `softmax` along a specific axis.

**Comment by benjamin-robbins** [comment]
How would that help with the need for a weighting aspect of the pixelwise classes?

**Comment by rex-yue-wu** [comment]
OK. My bad. All you need is to set `class weight` in `fit` or `fit_generator`. More specifically, you need to feed in a dictionary `class_weight = { 0:1, 1:your_weight }`, meaning that you set class-0 of weight 1 and class-1 of weight `your_weight`. If you have multiple classes, simply expand this dictionary. If you keras does not support this feature, please update it.

**Comment by benjamin-robbins** [comment]
Yeah it does, but I thought that the class_weights feature doesn't work pixelwise segmentation problems. see: [https://github.com/fchollet/keras/issues/3653](url) I have not been successful in getting sample_weights to work correctly either.

**Comment by vpoire** [comment]
I am also using u-net with a multiclass problem (nb_class > 2) that are not balanced. I computed the weight map in order to balance the classes. I ended up passing the weight map by adding an extra channel to y_true (nb_class+1). My network converge to a solution with Adam & RMSProp, but not with SGD. 

```
def loss(y_true, y_pred):
    epsilon = _to_tensor(_EPSILON, y_pred.dtype.base_dtype)
    y_pred = tf.clip_by_value(y_pred, epsilon, 1. - epsilon)
    return - tf.reduce_sum(tf.multiply(y_true[...,:nb_class] * tf.log(y_pred), y_true[...,nb_class]))

return loss
```
Hopwever, the loss curve also do weird things after a while. I am still trying to understand why I got such a big gradient magnitude, even with a very small learning rate (1e-12).

**Comment by wilkluka** [comment]
one idea is when you have neural network input called `input_x` to add two additional inputs `input_y` and `input_y_weights` and when you have variable with last layer output and let's call this variable `output` then you can add additional layer at the end:
```
# here we add 3 steps to perform loss-like behavior
difference = keras.layers.substract([outputs, input_y])
mse_loss = keras.layers.multiply([difference])
weighted_mse_loss = keras.layers.multiply([mse, input_y_weights])

model.compile(inputs= [input_x, input_y, input_y_weights], output = [weighted_mse_loss], loss='mae')

model.fit(x=inputs, y=np.zeros(inputs.shape))
```
and now last steps will compute your_custom_weighted_loss and loss mae on end will be for optimizer, don't forget that on prediction you should give 3 inputs 
`[actual_x, np.zeros(actual_x.shape), np.zeros(actual_x.shape)]`


whole trick is to compute loss in the network (it's like employee doing work for the boss that "don't have time for such low level jobs") :)

**Comment by m090009** [comment]
@benjamin-robbins  I'm curious to know how did you ended up solving the problem
